url4aac7e6f55_simon0191_tetris3d_tgz-pJ8-Tetris3dJ8/src/vsdk/toolkit/io/image/ImagePersistence.java

```diff
@@ -0,0 +1,693 @@
+        RGBPixel p;
+
+        xSize = img.getXSize();
+        ySize = img.getYSize();
+        bimg =  new BufferedImage(xSize, ySize, BufferedImage.TYPE_INT_RGB);
+        for ( y = 0; y < ySize; y++ ) {
+            for ( x = 0; x < xSize; x++ ) {
+                p = img.getPixelRgb(x, y);
+                bimg.setRGB(x, y, 
+                  (VSDK.signedByte2unsignedInteger(p.r)) * 256 * 256 +
+                  (VSDK.signedByte2unsignedInteger(p.g)) * 256 +
+                  (VSDK.signedByte2unsignedInteger(p.b))
+                );
+            }
+        }
+
+        ImageIO.write(bimg, "png", os);
+    }
+
+    public static void exportPNG(OutputStream os, Image img)
+        throws Exception
+    {
+/*
+        if ( img instanceof IndexedColorImage ) {
+//NOT WORKING
+            exportPNG_8bitGrayscale(os, (IndexedColorImage)img);
+        }
+        else {
+*/
+        exportPNG_24bitRgb(os, img);
+    }
+
+    /**
+    This method writes the contents of the specified image to a file in 
+    binary GIF image format. Returns true if everything
+    works fine, false if something fails, like a permission access denied
+    or if storage device runs out of space.
+    */    
+    public static boolean exportGIF(File fd, Image img)
+    {
+        try {
+            BufferedImage bimg;
+            int x, y, xSize, ySize;
+            RGBPixel p;
+
+            xSize = img.getXSize();
+            ySize = img.getYSize();
+            bimg =  new BufferedImage(xSize, ySize, BufferedImage.TYPE_INT_RGB);
+            for ( y = 0; y < ySize; y++ ) {
+                for ( x = 0; x < xSize; x++ ) {
+                    p = img.getPixelRgb(x, y);
+                    bimg.setRGB(x, y, 
+                      (VSDK.signedByte2unsignedInteger(p.r)) * 256 * 256 +
+                      (VSDK.signedByte2unsignedInteger(p.g)) * 256 +
+                      (VSDK.signedByte2unsignedInteger(p.b))
+                    );
+                }
+            }
+
+            FileOutputStream fos = new FileOutputStream(fd);
+
+            ImageIO.write(bimg, "gif", fos);
+
+            // OLD DEPRECATED API, DO NOT USE!
+            //com.sun.image.codec.jpeg.JPEGImageEncoder jpeg;
+            //jpeg = com.sun.image.codec.jpeg.JPEGCodec.createJPEGEncoder(fos);
+            //jpeg.encode(bimg);
+
+            fos.close();
+        }
+        catch ( Exception e ) {
+            return false;
+        }
+        return true;
+    }
+
+    /**
+    This method writes the contents of the specified image to a file in 
+    binary RGB PPM format (i.e. P6 PPM sub-format). Returns true if everything
+    works fine, false if something fails, like a permission access denied
+    or if storage device runs out of space.
+    */
+    public static boolean exportPPM(File fd, Image img)
+    {
+        try {
+            BufferedOutputStream writer;
+            FileOutputStream fos = new FileOutputStream(fd);
+            writer = new BufferedOutputStream(fos);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *writer*. The resource is referenced at lines 463, 472, 474, 476, 478, 485, 486, 487, 491. The resource is closed at lines 492, 493. There are other execution paths that do not contain closure statements, for example, when BufferedOutputStream.write throws an exception. Ensure *writer* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).